cmd/zc_traverser_blobfs_account.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2019 Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"path/filepath"
+
+	"github.com/Azure/azure-pipeline-go/pipeline"
+
+	"github.com/Azure/azure-storage-azcopy/azbfs"
+)
+
+// We don't allow S2S from BlobFS, but what this gives us is the ability for users to download entire accounts at once.
+// This is just added to create that feature parity.
+type BlobFSAccountTraverser struct {
+	accountURL        azbfs.ServiceURL
+	p                 pipeline.Pipeline
+	ctx               context.Context
+	fileSystemPattern string
+
+	// a generic function to notify that a new stored object has been enumerated
+	incrementEnumerationCounter func()
+}
+
+func (t *BlobFSAccountTraverser) traverse(processor objectProcessor, filters []objectFilter) error {
+	marker := ""
+	for {
+		resp, err := t.accountURL.ListFilesystemsSegment(t.ctx, &marker)
+
+		if err != nil {
+			return err
+		}
+
+		for _, v := range resp.Filesystems {
+			var fsName string
+			if v.Name != nil {
+				fsName = *v.Name
+			} else {
+				return errors.New("filesystem listing returned nil container names")
+			}
+
+			if t.fileSystemPattern != "" {
+				if ok, err := filepath.Match(t.fileSystemPattern, fsName); err != nil {
+					return err
+				} else if !ok {
+					continue
+				}
+			}
+
+			fileSystemURL := t.accountURL.NewFileSystemURL(fsName).URL()
+			fileSystemTraverser := newBlobFSTraverser(&fileSystemURL, t.p, t.ctx, true, t.incrementEnumerationCounter)
+
+			middlemanProcessor := initAccountMiddlemanProcessor(fsName, processor)
+
+			err = fileSystemTraverser.traverse(middlemanProcessor, filters)
+
+			if err != nil {
+				return err
+			}
+		}
+
+		if marker == "" {
+			break
```

Comment:
I notice that this marker handling code contains no `marker := nextMarker`.  This implies that either we have a bug, or the way that `ListFilesystemsSegment` handles the marker is a lot cleverer than what the blob code does.  I just took a quick look.  I don't think it is that clever. (It looks like it might be, because the type is a pointer, but I don't believe it ever updates it).

Reply:
So, azbfs doesn't have a marker struct.

Instead, it just uses a ptr to a string.

Reply:
All the marker struct _is_ anyway is just a pointer to a string. Which is odd because it just allocates a new string, and thus we have to update the marker in our code.

Reply:
Should we change this to be consistent with our other SDKs? I do feel like just using a struct that contains a pointer to another, new string just walks around what's _actually_ happening.

Reply:
I wouldn't bother making the BlobFS one consistent, because it will be phased out mostly (or entirely) once interop goes GA.  But I do suspect there's a bug in the current zc_traverser_blobfs_account, because I don't think anything is updating the marker to the next marker.  Our code isn't and, contrary to the impression given by the API's use of a pointer, I don't believe the azbfs code is updating it either.

Reply:
I'll take a look.

Reply:
Turns out this bug was in theory present in both the container and account traversers.